Handle thread interruption by closing the channel

lutovich opened the pull request:
Blocking API like `Session#run()` is now implemented on top of async API, i.e. `Session#runAsync()`. Blocking calls simply invoke async ones and uninterruptibly wait for them to complete. Previously, with blocking IO, thread interruptions resulted in `ClosedByInterruptException` which terminated query/transaction in a harsh way. Right now interrupts will do nothing. Handling them is not strictly required because `Thread#interrupt()` is not a nice thing to do. However, interrupts are still used in some environments as a last resort thing.

This commit makes driver react on thread interrupts by closing the underlying Netty channel. It's only done in places where channel is actually available. Closing of channel results in exception and the waiting thread fails. It will also make database reset state for this connection, terminate transaction and query.

Such approach seems very similar to the previous one except it will result in `ServiceUnavailableException` instead of `ClosedByInterruptException`.

lutovich committed:
Handle thread interruption by closing the channel

Blocking API like `Session#run()` is now implemented on top of async
API, i.e. `Session#runAsync()`. Blocking calls simply invoke async
ones and uninterruptibly wait for them to complete. Previously, with
blocking IO, thread interruptions resulted in
`ClosedByInterruptException` which terminated query/transaction in a
harsh way. Right now interrupts will do nothing. Handling them is not
strictly required because `Thread#interrupt()` is not a nice thing
to do. However, interrupts are still used in some environments as a
last resort thing.

This commit makes driver react on thread interrupts by closing the
underlying Netty channel. It's only done in places where channel is
actually available. Closing of channel results in exception and
the waiting thread fails. It will also make database reset state
for this connection, terminate transaction and query.

Such aproach seems very similar to the previous one except it will
result in `ServiceUnavailableException` instead of
`ClosedByInterruptException`.

ali-ince reviewed (approved):
Changes look good to me!

lutovich committed:
More tests for thread interruption handling

Added couple more integrations tests to verify that `Thread#interrupt()`
terminates active connection and "unblocks" the blocking thread:
  * when driver is created towards unresponsive server
  * when transaction blocks to read back the result
  * when transaction blocks to commit

Also `NettyConnection` will now fail provided handlers except throwing
exception directly. Increased test coverage for future helpers.

lutovich commented:
Added a commit with some more tests. @ali-ince could you please take a look?

ali-ince commented on driver/src/test/java/org/neo4j/driver/v1/GraphDatabaseTest.java:
```diff
@@ -156,4 +158,28 @@ public void shouldLogWhenUnableToCreateRoutingDriver() throws Exception
+            }
+            catch ( Exception ignore )
+            {
+                // expected
```
@lutovich Would it be better to expect a `ServiceUnavailableException` here instead of a `IllegalStateException`?

lutovich replied:
Yeah, that's true. What do you think about always wrapping exceptions from connectivity verification in `ServiceUnavailableException`? (unless it's already a `ServiceUnavailableException`)

lutovich committed:
Throw ServiceUnavailableException when unable to connect

Driver verifies connectivity on creation. It closes the connection pool
when thread is interrupted. Previously original `IllegalStateException`
from the closed pool has been propagated to the user.

This commit makes connectivity verification always throw
`ServiceUnavailableException`.

lutovich commented:
@ali-ince exceptions should now be fixed

lutovich committed:
Fixed couple errors after rebase

Also added unit tests for `Futures#asCompletionException()`.
